[GH-130] Adds a reminder when subscribing to a project that a web hook is needed.

ericjaystevens opened the pull request:
#### Summary
Adds a reminder when using ```/github subscribe``` that a project hook is needed if one is not already pointing to the MatterMosts server URL.

Webhooks in Gitlab can be created in GitLab at the System, project, or group level (paid plans only). The go-gitlab module used in this project does not yet support listing group hooks ( [issue created](https://github.com/xanzy/go-gitlab/issues/776) ). System hooks won't be accessible for most developers. So we are just checking for project hooks, and if we are unsure if a webhook exists, providing a link to our documentation on how to set one up.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-plugin-gitlab/issues/130

gabrieljackson commented:
Thanks for working on this @ericjaystevens.

A few things before we start reviewing:
1. I believe the linked ticket URL may have been a typo, so please correct that if so.
2. Please squash/fixup/etc. your commits a bit so that we start with a cleaner git history.
3. If possible, please increase coverage a bit on tests to the codecov tests pass.

Let us know if you need any help with this!

ericjaystevens commented:
@gabrieljackson Thanks for the feedback on this pull request. 

I have updated the linked ticket URL, squashed my commits, and improved coverage.

gabrieljackson commented:
Thanks @ericjaystevens. Reviewers have been added.

mickmister commented on server/command.go:
```diff
@@ -240,3 +189,86 @@ func (p *Plugin) ExecuteCommand(_ *plugin.Context, args *model.CommandArgs) (*mo
+	}
+
+	// Create subscription
+	if subscribeErr := p.Subscribe(info, namespace, project, channelID, features); err != nil {
```
I think `subscribeErr` is meant to be used here
```suggestion
	if subscribeErr := p.Subscribe(info, namespace, project, channelID, features); subscribeErr != nil {
```

mickmister commented on server/command.go:
```diff
@@ -240,3 +189,86 @@ func (p *Plugin) ExecuteCommand(_ *plugin.Context, args *model.CommandArgs) (*mo
+			"project", project,
+			"err", subscribeErr.Error(),
+		)
+		return err.Error()
```
```suggestion
		return subscribeErr.Error()
```

mickmister commented on server/gitlab/api.go:
```diff
@@ -11,6 +11,19 @@ import (
 
+// GetProjectHooks gathers all the project level hooks from a single GitLab project.
+func (g *gitlab) GetProjectHooks(user *GitlabUserInfo, owner string, repo string) ([]*internGitlab.ProjectHook, error) {
+
```
The code convention in the team's projects is usually no blank line at the beginning of the function. @ericjaystevens Do you have any objections on removing the blank lines?

mickmister replied:
@ericjaystevens Do you mind applying this fix to every function introduced in the PR?

mickmister commented on server/command.go:
```diff
@@ -240,3 +189,86 @@ func (p *Plugin) ExecuteCommand(_ *plugin.Context, args *model.CommandArgs) (*mo
 }
+
+// SubscribeCommand proccess the /gitlab subscribe command.
+// It returns a message and handles all errors my including helpfull information in the message
```
Nit for spelling
```suggestion
// It returns a message and handles all errors my including helpful information in the message
```

ericjaystevens commented on go.mod:
```diff
@@ -6,6 +6,7 @@ require (
 	github.com/mattermost/mattermost-server/v5 v5.18.0
 	github.com/pkg/errors v0.8.1
 	github.com/stretchr/testify v1.4.0
+	github.com/vektra/mockery v0.0.0-20181123154057-e78b021dcbb5 // indirect
```
I can change to go-mock no problem. I picked mockery because I saw it was used in [Mattermost-server](https://github.com/mattermost/mattermost-server/blob/7035e09fe9bbfd981bf73990d7857535f7b503aa/Makefile#L211). Can you expand on when go-mock is preferred?

ericjaystevens replied:
I have changed to go-mock

ericjaystevens commented:
@mickmister I think I addressed all the issues based on your great feedback. Let me know if you need anything else.

hanzei commented on go.mod:
```diff
@@ -1,11 +1,13 @@
 	github.com/mattermost/mattermost-server/v5 v5.18.0
 	github.com/pkg/errors v0.8.1
 	github.com/stretchr/testify v1.4.0
+	github.com/vektra/mockery v0.0.0-20181123154057-e78b021dcbb5 // indirect
```
Is this line needed any longer?

ericjaystevens replied:
It should be removed, I'll take it out over the weekend.

hanzei commented on go.sum:
```diff
@@ -190,15 +193,20 @@ github.com/magiconair/properties v1.8.0/go.mod h1:PppfXfuXeibc/6YijjN8zIbojt8czP
 github.com/magiconair/properties v1.8.1/go.mod h1:PppfXfuXeibc/6YijjN8zIbojt8czPbwD3XqdrwzmxQ=
 github.com/mailru/easyjson v0.0.0-20180730094502-03f2033d19d5/go.mod h1:C1wdFJiN94OJF2b5HbByQZoLdCWB1Yqtg26g4irojpc=
 github.com/mailru/easyjson v0.7.0/go.mod h1:KAzv3t3aY1NaHWoQz1+4F1ccyAH66Jk7yos7ldAVICs=
+github.com/manland/mattermost-plugin-gitlab v0.3.0 h1:4o5wOvLvdbez918JQz0ZMOTOLA3QDy4BEv2cFHWMeDk=
```
Does `github.com/manland/mattermost-plugin-gitlab` still need to be imported?

ericjaystevens replied:
I expect this can be removed too.

ericjaystevens commented:
Removed all extra lines at the beginning of functions and cleaned up go.sum and go.mod. Let me know if there is anything else I can do. Thanks again for all the feedback to improve this PR.

ericjaystevens commented:
@DHaussermann  Good Catch the URL must have recently changed. Thanks! I just pushed with the updated link. Let me know if you need anything else.

I forgot to update the URL's in the tests in my last commit, just pushed those changes up.
